sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/implementation/SessionNotAvailableRetryTest.java

```diff
@@ -0,0 +1,338 @@
+            ReflectionUtils.getGlobalEndpointManager(rxDocumentClient);
+        DatabaseAccount databaseAccount = globalEndpointManager.getLatestDatabaseAccount();
+        this.databaseAccount = databaseAccount;
+
+    }
+
+    @DataProvider(name = "preferredRegions")
+    private Object[][] preferredRegions() {
+        List<String> preferredLocations1 = new ArrayList<>();
+        List<String> regionalSuffix1 = new ArrayList<>();
+        List<String> preferredLocations2 = new ArrayList<>();
+        List<String> regionalSuffix2 = new ArrayList<>();
+        Iterator<DatabaseAccountLocation> locationIterator = this.databaseAccount.getReadableLocations().iterator();
+        while (locationIterator.hasNext()) {
+            DatabaseAccountLocation accountLocation = locationIterator.next();
+            preferredLocations1.add(accountLocation.getName());
+            regionalSuffix1.add(getRegionalSuffix(accountLocation.getEndpoint(), TestConfigurations.HOST));
+        }
+
+        //putting preference in opposite direction than what came from database account api
+        for (int i = preferredLocations1.size() - 1; i >= 0; i--) {
+            preferredLocations2.add(preferredLocations1.get(i));
+            regionalSuffix2.add(regionalSuffix1.get(i));
+        }
+
+        return new Object[][]{
+            new Object[]{preferredLocations1, regionalSuffix1, OperationType.Read},
+            new Object[]{preferredLocations2, regionalSuffix2, OperationType.Read},
+            new Object[]{preferredLocations1, regionalSuffix1, OperationType.Query},
+            new Object[]{preferredLocations2, regionalSuffix2, OperationType.Query},
+            new Object[]{preferredLocations1, regionalSuffix1, OperationType.Create},
+            new Object[]{preferredLocations2, regionalSuffix2, OperationType.Create},
+        };
+    }
+
+    @Test(groups = {"multi-master"}, dataProvider = "preferredRegions")
+    public void sessionNotAvailableRetryMultiMaster(List<String> preferredLocations, List<String> regionalSuffix,
+                                                    OperationType operationType) throws IllegalAccessException {
+        CosmosAsyncClient preferredListClient = new CosmosClientBuilder()
+            .endpoint(TestConfigurations.HOST)
+            .key(TestConfigurations.MASTER_KEY)
+            .contentResponseOnWriteEnabled(true)
+            .directMode()
+            .preferredRegions(preferredLocations)
+            .buildAsyncClient();
+
+        AsyncDocumentClient asyncDocumentClient = (AsyncDocumentClient) FieldUtils.readField(preferredListClient,
+            "asyncDocumentClient", true);
+        RxDocumentClientImpl rxDocumentClient = (RxDocumentClientImpl) asyncDocumentClient;
+        ServerStoreModel storeModel = (ServerStoreModel) FieldUtils.readField(rxDocumentClient, "storeModel", true);
+        StoreClient storeClient = (StoreClient) FieldUtils.readField(storeModel, "storeClient", true);
```

Comment:
we have helper method for these in ReflectionUtils. please use that. for the cases we don't have helper method, please add the helper method in the ReflectionUtils.

same for all other usage of FieldUtils.readField, please try to rely on the ReflectionUtils which is meant for this.

Reply:
I am avoiding extra method creation in ReflectionUtils for both getter and setter if it is not generic enough to be use more than one place.

Reply:
I moved most of them in ReflectionUtils  , which will be reused in future

Reply:
Naveen, the intention behind ReflectionUtils is to centralized these kind of access in one place.

If in future we change the parameter name, e.g, storeModel to storemodel, with the centralized approach we will have to do it in one place rather than finding all tests and fixing it in all tests.